File: app_canibuythat/src/main/java/com/gb/canibuythat/model/BalanceCalculator.java
package com.gb.canibuythat.model;


import java.util.Calendar;
import java.util.Date;

import com.gb.canibuythat.util.DateUtils;


/**
 * Created by gbiro on 1/30/2015.
 */
public class BalanceCalculator {

	private BudgetModifier	budgetModifier;


	public BalanceCalculator(BudgetModifier budgetModifier) {
		this.budgetModifier = budgetModifier;
	}


	/**
	 * See how many times the specified <code>budgetModifier</code> has been applied in the past (up until today) and
	 * sum up it's value.
	 *
	 * @return two floating point values, the first is the minimum value (because there are periods where the
	 *         application of the modifier is uncertain), the second is the maximum value
	 */
	public float[] getEstimatedBalance(BudgetModifier budgetModifier, Date startingFrom) {
		boolean exit = false;
		float minimum = 0;
		float maximum = 0;
		int count = 0;
		Calendar c1 = Calendar.getInstance();
		c1.setTime(budgetModifier.lowerDate);
		Calendar c2 = Calendar.getInstance();
		c2.setTime(budgetModifier.upperDate);
		Calendar startDate = Calendar.getInstance();

		if (startingFrom != null) {
			startDate.setTime(startingFrom);
		}
		startDate = DateUtils.clearLowerBits(startDate);

		do {
			if (areWeBeforeBMStart(startDate, c1)) {
				exit = true;
			} else if (areWeWithinBMPeriod(startDate, c1, c2)) {
				minimum += budgetModifier.amount * budgetModifier.type.sign;
			} else if (areWeAfterBmEnd(startDate, c2)) {
				minimum += budgetModifier.amount * budgetModifier.type.sign;
				maximum += budgetModifier.amount * budgetModifier.type.sign;
			}
			increaseDateWithPeriod(c1, budgetModifier.periodType, budgetModifier.periodMultiplier);
			increaseDateWithPeriod(c2, budgetModifier.periodType, budgetModifier.periodMultiplier);
			count++;

			if (budgetModifier.repetitionCount != null && count >= budgetModifier.repetitionCount) {
				exit = true;
			}
		} while (!exit);
		return new float[] {
				minimum, maximum
		};
	}


	boolean areWeBeforeBMStart(Calendar now, Calendar lowerDate) {
		return now.before(lowerDate);
	}


	/**
	 * Boundaries included
	 */
	boolean areWeWithinBMPeriod(Calendar now, Calendar lowerDate, Calendar upperDate) {
		return now.getTimeInMillis() >= lowerDate.getTimeInMillis()
				&& now.getTimeInMillis() <= upperDate.getTimeInMillis();
	}


	boolean areWeAfterBmEnd(Calendar now, Calendar upperDate) {
		return now.getTimeInMillis() > upperDate.getTimeInMillis();
	}


	public static void increaseDateWithPeriod(Calendar c, BudgetModifier.PeriodType period, int periodMultiplier) {
		switch (period) {
		case DAYS:
			c.add(Calendar.DAY_OF_MONTH, periodMultiplier);
			break;
		case WEEKS:
			c.add(Calendar.WEEK_OF_MONTH, periodMultiplier);
			break;
		case MONTHS:
			c.add(Calendar.MONTH, periodMultiplier);
			break;
		case YEARS:
			c.add(Calendar.YEAR, periodMultiplier);
			break;
		}
	}
}

File: app_canibuythat/src/main/java/com/gb/canibuythat/repository/BalanceCalculator.kt
package com.gb.canibuythat.repository

import com.gb.canibuythat.model.*
import java.time.LocalDate

object BalanceCalculator {

    /**
     * Calculate how many times the specified `spending` has been
     * applied up until the specified `end` date (usually today) and sum up
     * it's value.

     * @param end if null, `today` is used instead
     * *
     * @return two floating point values, the first is the minimum possible value, the second is
     * * the maximum possible value
     */
    fun getEstimatedBalance(spending: Spending, start: LocalDate?, end: LocalDate = LocalDate.now()): Balance {
        if (start != null && end < start) {
            throw IllegalArgumentException("Start date must come before end date!")
        }
        var `break` = false
        var definitely = 0f
        var maybe = 0f
        var targetDefinitely = 0f
        var targetMaybe = 0f
        var balance = 0f
        var targetBalance = 0f
        var occurrenceCount = 0
        val spendingEvents = mutableListOf<SpendingEvent>()
        var movingStart = spending.fromStartDate
        var movingEnd = spending.fromEndDate
        var counter = 0
        do {
            if (start == null || movingEnd >= start) {
                val target = spending.target?.let { if (it > 0) -it.toFloat() else it.toFloat() }
                        ?: spending.value.toFloat()
                if (end >= movingStart) {
                    if (spending.enabled) {
                        maybe += spending.value.toFloat()
                        targetMaybe += target
                    }
                    if (end > movingEnd) {
                        if (spending.enabled) {
                            definitely += spending.value.toFloat()
                            targetDefinitely += target
                            balance += spending.value.toFloat()
                            targetBalance += target
                        }
                    }
                    spendingEvents.add(SpendingEvent(movingStart, movingEnd, definitely, maybe))
                } else {
                    `break` = true
                }
            }
            counter++
            movingStart = spending.fromStartDate + (spending.cycleMultiplier * counter * spending.cycle)
            movingEnd = spending.fromEndDate + (spending.cycleMultiplier * counter * spending.cycle)
            spending.occurrenceCount?.let {
                if (++occurrenceCount >= it) {
                    `break` = true
                }
            }
        } while (!`break`)
        return Balance(definitely, maybe, balance, targetDefinitely, targetMaybe, targetBalance, spendingEvents.toTypedArray())
    }
}
